File: 2015/05/5.ts
import { $, filter, length, lines, map, readInput } from '../../common'
const input = $(readInput(), lines)

const part1niceRegex1 = /[aeiou].*[aeiou].*[aeiou]/
const part1niceRegex2 = /(.)\1/
const part1naughtyRegex = /ab|cd|pq|xy/

const part1isNice = (s: string) => part1niceRegex1.test(s) && part1niceRegex2.test(s) && !part1naughtyRegex.test(s)

console.log('Part 1:', $(input, map(part1isNice), filter(Boolean), length))

const part2niceRegex1 = /(..).*\1/
const part2niceRegex2 = /(.).\1/

const part2isNice = (s: string) => part2niceRegex1.test(s) && part2niceRegex2.test(s)

console.log('Part 2:', $(input, map(part2isNice), filter(Boolean), length))

File: 2025/05/5.clj
(ns aoc.2025.05.5
  (:require
   [aoc.common :refer [lines parse-input]]
   [blancas.kern.core :refer [<$> <*> << dec-num many-till new-line* sep-by sym*]]))

(defn in-range [ranges ingredient]
  (->> ranges
       (some (fn [[from to]] (<= from ingredient to)))))

(defn shrink-ranges [[first & rest]]
  (->> rest
       (reduce (fn [ranges [start end]]
                 (let [[_ test-end] (peek ranges)]
                   (cond
                     (<= end test-end)  ranges                                ; full overlapp
                     (> start test-end) (conj ranges [start end])             ; null overlapp
                     :else              (conj ranges [(inc test-end) end])))) ; litt overlapp
               [first])))

(let [[ranges ingredients] (parse-input
                            (<*> (<$> (partial sort-by first)
                                      (many-till (<< (sep-by (sym* \-) dec-num) new-line*) new-line*))
                                 (lines dec-num)))]
  (->> ingredients
       (filter (partial in-range ranges))
       count
       (println "Part 1:"))

  (->> ranges
       shrink-ranges
       (map #(inc (abs (apply - %1))))
       (apply +)
       (println "Part 2:")))